feat(buildimages)!:Adapt tasks to handle new images

chouetz opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
Update the `buildimages` tasks to be able to handle the new `CI_IMAGE_` buildimage variables. I also introduced these new images (not used yet) to prepare migration

### Motivation
Smart build of buildimage, capacity to build & promote a set of images to `datadog-agent`

### Describe how to test/QA your changes
UT + local manual tests

### Possible Drawbacks / Trade-offs
This is a breaking change in the `buildimages.update` task.
I also decided to change the `test_version` argument to `test` and aligned everywhere, so increased the impact of this PR

### Additional Notes
Based on previous changes, #30365 and #30370
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

chouetz committed:
Add new variables on gitlab configuration

chouetz committed:
feat(buildimages)!:Adapt tasks to handle new images

chouetz committed:
fix tests

chouetz committed:
manual tests

chouetz committed:
align test argument everywhere

CelianR commented on tasks/buildimages.py:
```diff
@@ -82,19 +87,36 @@ def generate_pr_body(
+    if image_type is None:
+        return gitlab_ci["variables"].get(
+            "DATADOG_AGENT_BUILDIMAGES",
+            f'{color_message("Not found", "red")}: DATAOG_AGENT_BUILDIMAGES is not defined in the configuration',
```
```suggestion
            f'{color_message("Not found", "red")}: DATADOG_AGENT_BUILDIMAGES is not defined in the configuration',
```

spencergilbert commented on tasks/buildimages.py:
```diff
@@ -2,24 +2,32 @@
-        "image_tag": "tag from build_image with format v<build_id>_<commit_id>",
-        "test_version": "Is a test image or not",
+        "tag": "tag from build_image with format v<build_id>_<commit_id>",
+        "images": "The image(s) to update, comma separated. If empty, updates the DATADOG_AGENT_ buildimages. Can be a incomplete pattern, e.g. 'deb,rpm' will update all deb and rpm images. deb_x64 will update only one image. Use `all` to update all images",
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏ Is there an exhaustive list to make it easier to reference possible patterns? Just the `gitlab-ci.yml` variables section?

chouetz replied:
Yes just this. Maybe I could add an option to list possible values.

spencergilbert replied:
Something to think about - I don't know if I'd have a good idea of _what_ I could provide here that isn't "all" but if we're not usually needing a subset, it may als obe fine

chouetz replied:
Added an option to list values, with `--list-images` on [this commit](https://github.com/DataDog/datadog-agent/pull/30380/commits/aff1a68aefe20cdddc44f711a5c1e851b7549878)

chouetz committed:
fix unit test after change of the variable name

chouetz commented on tasks/libs/ciproviders/gitlab_api.py:
```diff
@@ -1250,3 +1250,71 @@ def full_config_get_all_stages(full_config: dict) -> set[str]:
+    return variables_to_update
+
+
+def filter_variables(variables, images):
```
I thought the naming convention was here to indicate what the variables are about, when you are looking or editing the file. In the context of invoke task execution I think what matters is only the image type, so I thought we could get rid of the prefixes

chouetz replied:
I finally did your suggestion, there was no point in removing prefix and suffix

chouetz committed:
Use removesuffix instead of replace

Co-authored-by: Kevin Fairise <132568982+KevinFairise2@users.noreply.github.com>

pgimalac commented on tasks/buildimages.py:
```diff
@@ -82,19 +87,36 @@ def generate_pr_body(
 @task(
     help={
         "file_path": "path of the Gitlab configuration YAML file",
+        "image_type": "The type of image to get the tag for. Used for CI_IMAGE_ variables",
```
What can this argument be set to ?

chouetz replied:
As you have understood, there is no `choices` predefined, but it's maybe not that secure, I'll fix this.

chouetz replied:
I updated this method to only accept valid image names. The list can be obtained with the `inv buildimages.update --list-image` task

pgimalac commented on tasks/buildimages.py:
```diff
@@ -82,19 +87,36 @@ def generate_pr_body(
+        for key in gitlab_ci["variables"].keys():
+            if key.startswith("CI_IMAGE"):
+                available_images.add(key.removeprefix("CI_IMAGE_").removesuffix("_SUFFIX").casefold())
+                if image_type in key.casefold():
```
Why `in` ? Why not just equals ?

pgimalac replied:
If I understand correctly if someone used `deb` then they would get the version of the first `deb` image listed, but each image could have a different version so you can't know which one you'll get a priori

chouetz replied:
Yes for this tasks it's not useful

pgimalac commented on .gitlab-ci.yml:
```diff
@@ -181,6 +181,41 @@ variables:
+  CI_IMAGE_DOCKER_X64_SUFFIX: ""
+  CI_IMAGE_DOCKER_ARM64: v46542806-c7a4a6be
+  CI_IMAGE_DOCKER_ARM64_SUFFIX: ""
+  CI_IMAGE_GILBC_2_17_X64: v46542806-c7a4a6be
+  CI_IMAGE_GILBC_2_17_X64_SUFFIX: ""
+  CI_IMAGE_GILBC_2_23_ARM64: v46542806-c7a4a6be
+  CI_IMAGE_GILBC_2_23_ARM64_SUFFIX: ""
```
I guess ?
```suggestion
  CI_IMAGE_GLIBC_2_17_X64: v46542806-c7a4a6be
  CI_IMAGE_GLIBC_2_17_X64_SUFFIX: ""
  CI_IMAGE_GLIBC_2_23_ARM64: v46542806-c7a4a6be
  CI_IMAGE_GLIBC_2_23_ARM64_SUFFIX: ""
```

chouetz replied:
Wow nice catch, thanks

pgimalac commented on tasks/buildimages.py:
```diff
@@ -12,17 +12,22 @@
-        "image_tag": "tag from build_image with format v<build_id>_<commit_id>",
-        "test_version": "Is a test image or not",
+        "tag": "tag from build_image with format v<build_id>_<commit_id>",
+        "images": "The image(s) to update, comma separated. If empty, updates the DATADOG_AGENT_ buildimages. Can be a incomplete pattern, e.g. 'deb,rpm' will update all deb and rpm images. deb_x64 will update only one image. Use `all` to update all images",
```
I feel like `images` being a substring of the image names makes it super easy to change too many images...
If this will be called automatically by a script or by CI, then I expect it will explicitly list every image changed, not using that logic
Is there a case where you expect someone to use `deb` for example ? It would mean you changed all deb images and no other, which seems tricky to ensure

chouetz replied:
Yes it was for a human to make it simpler to call when, for instance, you modify all `deb` or all `x64`. As you say we can consider
* if done automatically then the full list can be given
* buildimages can provide this list through the notification system
So I can change the logic to make it more strict (and then probably less error prone)

pgimalac reviewed (approved):
LGTM for ASC owned files, however I have a hard time understanding the changes in other files (image selection logic), could you add comments in that code to explain what it does ?

chouetz committed:
codereview: several fixes

- Force to set the "real" image name in the `get_tag` task
- Set the new CI_IMAGE variables to the same names used in buildimages
- Change the `update_gitlab_config` method to always update the
  DATADOG_AGENT_ variables

chouetz committed:
Add a new option to list images

chouetz committed:
renamed methods, update comments and added a test

chouetz committed:
Update comment

chouetz commented:
/merge

chouetz commented:
/merge
